### ISSUE TITLE
The locality hypothesis

### ISSUE THREAD

**Issue by celoyd** [celoyd/potato#18]
I believe, but cannot prove, that pansharpening can be done well with a small receptive field. Chatted about this with @meetar and realized I should write it down, because it’s a load-bearing assumption in the model design.

Clarifying the claim without trying to be too rigorous: Putting aside all practicalities, imagine the ideal pansharpeners – those that give the best possible outputs that could possibly be estimated from their inputs, given a particular input size. You can even envision highly skilled humans with infinite time and a lot of paint. I’m arguing that the ones with small receptive fields, where only a small radius of context around a given pixel can affect what it looks like in the output, are almost as good as the ones with big receptive fields.

![image](https://github.com/celoyd/pansharpening/assets/148985/385975e7-16b9-405c-9754-b3381b95213e)

It doesn’t matter how we’re measuring quality or whether the field is a radius or a diameter in pan or multispectral pixels; my argument is only that the curve is steep at first and soon becomes almost flat.

This is based on a few observations and intuitions, including:

* If you imagine looking at an image patch, the amount by which your estimate of color distribution near the center of the patch changes as the patch grows will fall off very quickly.
* Actual pansharpening output quality per cycle (including in training time if applicable) varies between real-world models in a way that doesn’t seem to depend heavily on patch size.
* In fact, algorithms that only consider upsampled multispectral and panchromatic data on a single-pixel basis can perform _reasonably_. (These have the receptive field of the upsampling method.)

This is why our current model has a quite small receptive field, somewhere around 18 multispectral pixels last I checked. It’s why I’m not particularly interested in QKV transformers – clearly they _can_ work for this kind of task, but I’m not convinced that they’re worth the conceptual or [computational](https://arxiv.org/abs/2209.04881) complexity or the slow training.

I’m not claiming that the curve actually goes flat, or that a transformer is never worth it, or that I won’t try throwing an S4 module in a pansharpening model one day. It is certainly possible to imagine a scenario where very wide receptive fields help. 

An example I think about is a field of pumpkins. Each pumpkin is smaller than a multispectral pixel and thus a rich orange hue will probably never actually appear in the input; it’s always mixed with background colors. The ideal model still notices that there are a lot of pixels in the area reaching toward a rich orange endmember, and lets them “talk to each other” across space. Each pumpkin should radiate a _there are pumpkins nearby_ field. (Less colorfully, you could make similar arguments about the model gaining an understanding of, say, the band misalignment in an image neighborhood.) Wide receptive fields – probably of the transformer type – are clearly the answer to this problem. I admit this. But I think any pansharpener I can train at home is going to hit other practical limits before that one. I’m willing to have suboptimal pumpkin fields if optimal ones can’t even run on my old consumer graphics card.

In short, I think increasing the radius available to a pansharpening operator gives a sublinear increase in quality while requiring a superlinear increase in computation. Weighted in any sensible way, the curves don’t get very far before they cross.
